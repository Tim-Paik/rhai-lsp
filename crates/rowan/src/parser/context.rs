//! The parser context is a separate module to limit
//! the API surface for the parser functions.
#![allow(dead_code)]
use rowan::{GreenNodeBuilder, TextRange, TextSize};

use crate::syntax::{
    Lexer,
    SyntaxKind::{self, *},
};

use super::{Parse, ParseError, ParseErrorKind};

pub(crate) struct Context<'src> {
    lexer: Lexer<'src>,
    current_token: Option<SyntaxKind>,
    last_token: Option<SyntaxKind>,
    green: GreenNodeBuilder<'static>,
    errors: Vec<ParseError>,

    /// Tracks statements being separated by ";".
    statement_closed: bool,
}

impl<'src> Context<'src> {
    pub(crate) fn new(source: &'src str) -> Self {
        Self {
            lexer: Lexer::new(source),
            current_token: None,
            last_token: None,
            green: GreenNodeBuilder::new(),
            errors: Vec::new(),

            statement_closed: true,
        }
    }

    pub(crate) fn token(&mut self) -> Option<SyntaxKind> {
        // Eat insignificant tokens
        loop {
            if self.current_token.is_none() {
                self.current_token = self.lexer.next();
            }

            match self.current_token {
                Some(COMMENT_BLOCK) | Some(COMMENT_LINE) | Some(WHITESPACE) => {
                    self.eat();
                    self.last_token = None // We aren't interested in these.
                }
                Some(ERROR) => {
                    self.eat_error(ParseErrorKind::InvalidInput);
                    self.last_token = None // We aren't interested in these.
                }
                _ => break,
            }
        }

        self.current_token
    }

    pub(crate) fn previous_token(&self) -> Option<SyntaxKind> {
        self.last_token
    }

    pub(crate) fn eat(&mut self) {
        if let Some(t) = self.current_token.take() {
            self.green.token(t.into(), self.lexer.slice());
            self.last_token = Some(t);
        }
        self.current_token = None;
    }

    pub(crate) fn eat_as(&mut self, kind: SyntaxKind) {
        if self.current_token.is_some() {
            self.green.token(kind.into(), self.lexer.slice());
            self.last_token = Some(kind);
        }
        self.current_token = None;
    }

    pub(crate) fn eat_error(&mut self, error: ParseErrorKind) {
        self.add_error(error);
        self.eat();
    }

    pub(crate) fn eat_error_as(&mut self, kind: SyntaxKind, error: ParseErrorKind) {
        self.add_error(error);
        self.eat_as(kind);
    }

    pub(crate) fn add_error(&mut self, error: ParseErrorKind) {
        let span = self.lexer.span();
        self.errors.push(ParseError::new(
            TextRange::new(
                TextSize::from(span.start as u32),
                TextSize::from(span.end as u32),
            ),
            error,
        ));
    }

    pub(crate) fn start_node(&mut self, kind: SyntaxKind) {
        self.green.start_node(kind.into());
    }

    pub(crate) fn finish_node(&mut self) {
        self.green.finish_node();
    }

    pub(crate) fn finish(self) -> Parse {
        Parse {
            errors: self.errors,
            green: self.green.finish(),
        }
    }

    pub(crate) fn statement_closed(&self) -> bool {
        self.statement_closed
    }

    pub(crate) fn set_statement_closed(&mut self, v: bool) {
        self.statement_closed = v;
    }
}
