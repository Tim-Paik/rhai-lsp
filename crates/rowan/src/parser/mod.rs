//! This module contains all parsing-related tools including a
//! recursive-descent parser.

use crate::syntax::{SyntaxKind, SyntaxNode};
use rowan::GreenNode;
use thiserror::Error;

use self::context::Context;

mod context;
pub mod parsers;

/// A flexible parser.
/// 
/// Parsing happens via given parser functions (also found in [`parsers`]).
/// 
/// # Example Uses
/// 
/// ```
/// use rhai_rowan::parser::Parser;
/// // Parse some Rhai code (file).
/// let parser = Parser::new(r#"print("hello rhai!")"#);
/// let parse = parser.parse();
/// 
/// assert!(parse.errors.is_empty());
/// ```
/// 
/// ```
/// use rhai_rowan::parser::Parser;
/// use rhai_rowan::parser::parsers::parse_expr;
///
/// // Parse an expression explicitly.
/// let mut parser = Parser::new(r#"print("hello rhai!")"#);
/// parser.execute(parse_expr);
/// let parse = parser.finish();
/// 
/// assert!(parse.errors.is_empty());
/// ```
pub struct Parser<'src> {
    context: Context<'src>,
}

impl<'src> Parser<'src> {
    /// Create a new parser with the given source.
    #[must_use]
    pub fn new(source: &'src str) -> Self {
        Self {
            context: Context::new(source),
        }
    }

    /// Finish parsing.
    /// 
    /// # Panics
    /// 
    /// If no parser function was called,
    /// or a parser was left in an invalid state (bug).
    #[must_use]
    pub fn finish(self) -> Parse {
        self.context.finish()
    }
}

impl<'src> Parser<'src> {
    /// Execute a parser function for the given parser.
    pub fn execute<F: FnOnce(&mut Context)>(&mut self, f: F) {
        f(&mut self.context);
    }
}

/// The result of parsing, containing all errors
/// and the final green tree.
#[derive(Debug, Clone)]
pub struct Parse {
    /// Parse errors.
    pub errors: Vec<ParseError>,
    /// Parsed green tree.
    pub green: GreenNode,
}

impl Parse {
    /// Turn the result green tree into a CST.
    /// *This ignores errors*, the resulting tree
    /// can be potentially syntactically invalid.
    #[must_use]
    pub fn clone_syntax(&self) -> SyntaxNode {
        SyntaxNode::new_root(self.green.clone())
    }

    /// Turn the result green tree into a CST.
    /// *This ignores errors*, the resulting tree
    /// can be potentially syntactically invalid.
    #[must_use]
    pub fn into_syntax(self) -> SyntaxNode {
        SyntaxNode::new_root(self.green)
    }
}

/// A parse (syntax) error.
#[derive(Debug, Clone, PartialEq, Eq, Error)]
#[error("parse error at {range:?}: {kind}")]
pub struct ParseError {
    /// The span of the error in the parsed source.
    pub range: rowan::TextRange,
    /// Error kind.
    pub kind: ParseErrorKind,
}

impl ParseError {
    fn new(range: rowan::TextRange, kind: ParseErrorKind) -> Self {
        Self { range, kind }
    }
}

/// All the non-fatal parsing errors that can occur.
#[derive(Debug, PartialEq, Eq, Clone, Error)]
pub enum ParseErrorKind {
    #[error("unexpected EOF")]
    UnexpectedEof,

    #[error(r#"invalid input"#)]
    InvalidInput,

    #[error(r#"unexpected token"#)]
    UnexpectedToken,

    #[error(r#"expected token "{0:?}""#)]
    ExpectedToken(SyntaxKind),

    #[error(
        r#"expected one of "{}""#,
        .0
            .iter()
            .map(|s|format!("{:?}", s))
            .collect::<Vec<String>>().join(",")
    )]
    ExpectedOneOfTokens(Vec<SyntaxKind>),
}
