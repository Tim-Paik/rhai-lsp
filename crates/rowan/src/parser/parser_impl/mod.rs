#![deny(unreachable_patterns)]

use super::context::Context;
use crate::parser::ParseErrorKind;
use crate::syntax::SyntaxKind::*;
use crate::T;

macro_rules! require_token {
    ($ctx:ident) => {
        match $ctx.token() {
            Some(t) => t,
            None => return $ctx.eat_error(ParseErrorKind::UnexpectedEof),
        }
    };

    ($ctx:ident, finish_node) => {
        match $ctx.token() {
            Some(t) => t,
            None => {
                $ctx.finish_node();
                $ctx.eat_error(ParseErrorKind::UnexpectedEof);
                return;
            }
        }
    };
}

macro_rules! expect_token {
    ($ctx:ident, finish_node, $($token:tt)*) => {
        match $ctx.token() {
            Some(t) => match t {
                $($token)* => {
                    $ctx.eat();
                },
                _ => {
                    $ctx.finish_node();
                    $ctx.eat_error(ParseErrorKind::ExpectedToken($($token)*));
                    return;
                }
            }
            None => {
                $ctx.finish_node();
                $ctx.eat_error(ParseErrorKind::ExpectedToken($($token)*));
                return;
            }
        }
    };
    ($ctx:ident, $($token:tt)+) => {
        match $ctx.token() {
            Some(t) => match t {
                $($token)* => {
                    $ctx.eat();
                },
                _ => return $ctx.eat_error(ParseErrorKind::UnexpectedEof),
            }
            None => return $ctx.eat_error(ParseErrorKind::UnexpectedEof),
        }
    };

}

impl<'src> super::Parser<'src> {
    pub fn parse_file(&mut self) {
        parse_file(&mut self.context);
    }
}

fn parse_file(ctx: &mut Context) {
    ctx.start_node(FILE);
    if let Some(SHEBANG) = ctx.token() {
        parse_shebang(ctx);
    }

    ctx.set_statement_closed(true);
    while ctx.token().is_some() {
        if !ctx.statement_closed() {
            ctx.add_error(ParseErrorKind::ExpectedToken(T![";"]));
        }

        parse_stmt(ctx);
    }
    ctx.set_statement_closed(true);

    ctx.finish_node()
}

fn parse_shebang(ctx: &mut Context) {
    let token = require_token!(ctx);

    if !matches!(token, SHEBANG) {
        return ctx.eat_error(ParseErrorKind::ExpectedToken(SHEBANG));
    }

    ctx.eat()
}

fn parse_stmt(ctx: &mut Context) {
    let token = require_token!(ctx);
    ctx.start_node(STMT);
    ctx.set_statement_closed(false);

    if token == T![";"] {
        ctx.finish_node();
        ctx.set_statement_closed(true);
        return ctx.eat();
    }

    parse_item(ctx);

    if let Some(token) = ctx.token() {
        if token == T![";"] {
            ctx.eat();
            ctx.set_statement_closed(true);
        }
    }

    ctx.finish_node();
}

fn parse_item(ctx: &mut Context) {
    ctx.start_node(ITEM);

    // Parse doc comments if any.
    while matches!(
        require_token!(ctx, finish_node),
        COMMENT_BLOCK_DOC | COMMENT_LINE_DOC
    ) {
        ctx.eat();
    }

    parse_expr(ctx);

    ctx.finish_node();
}

fn parse_expr(ctx: &mut Context) {
    ctx.start_node(EXPR);
    let token = require_token!(ctx, finish_node);

    match token {
        T!["let"] => parse_expr_let(ctx),
        T!["const"] => parse_expr_const(ctx),
        T!["{"] => parse_expr_block(ctx),
        T!["fn"] => parse_expr_fn(ctx),
        LIT_INT | LIT_FLOAT | LIT_BOOL | LIT_STR | LIT_CHAR => parse_expr_lit(ctx),
        IDENT => parse_expr_ident(ctx),
        _ => ctx.eat_error(ParseErrorKind::UnexpectedToken),
    }

    ctx.finish_node();
}

fn parse_expr_ident(ctx: &mut Context) {
    ctx.start_node(EXPR_IDENT);
    parse_ident(ctx);
    ctx.finish_node();
}

fn parse_expr_lit(ctx: &mut Context) {
    ctx.start_node(EXPR_LIT);
    parse_lit(ctx);
    ctx.finish_node();
}

fn parse_expr_let(ctx: &mut Context) {
    ctx.start_node(EXPR_LET);

    expect_token!(ctx, finish_node, T!["let"]);
    expect_token!(ctx, finish_node, T!["ident"]);
    expect_token!(ctx, finish_node, T!["="]);

    parse_expr(ctx);

    ctx.finish_node();
}

fn parse_expr_const(ctx: &mut Context) {
    ctx.start_node(EXPR_CONST);

    expect_token!(ctx, finish_node, T!["const"]);
    expect_token!(ctx, finish_node, T!["ident"]);
    expect_token!(ctx, finish_node, T!["="]);

    parse_expr(ctx);

    ctx.finish_node();
}

fn parse_expr_block(ctx: &mut Context) {
    ctx.start_node(EXPR_BLOCK);

    expect_token!(ctx, finish_node, T!["{"]);

    ctx.set_statement_closed(true);
    loop {
        let token = require_token!(ctx, finish_node);

        if token == T!["}"] {
            break;
        }

        if !ctx.statement_closed() {
            ctx.add_error(ParseErrorKind::ExpectedToken(T![";"]));
        }

        parse_stmt(ctx);
    }

    expect_token!(ctx, finish_node, T!["}"]);

    // Blocks also don't require ";" when used as statements.
    ctx.set_statement_closed(true);

    ctx.finish_node();
}

fn parse_expr_fn(ctx: &mut Context) {
    ctx.start_node(EXPR_FN);

    expect_token!(ctx, finish_node, T!["fn"]);
    expect_token!(ctx, finish_node, T!["ident"]);

    parse_param_list(ctx);
    parse_expr_block(ctx);

    ctx.finish_node();
}

fn parse_param_list(ctx: &mut Context) {
    ctx.start_node(PARAM_LIST);

    let start_token = require_token!(ctx, finish_node);

    match start_token {
        T!["("] | T!["|"] => {
            ctx.eat();
        }
        _ => {
            ctx.eat_error(ParseErrorKind::ExpectedOneOfTokens(vec![T!["("], T!["|"]]));
            return;
        }
    }

    loop {
        let token = require_token!(ctx, finish_node);
        if matches!(
            (start_token, token),
            (T!["("], T![")"]) | (T!["|"], T!["|"])
        ) {
            ctx.eat();
            break;
        }

        parse_param(ctx);

        let end_token = require_token!(ctx, finish_node);

        match (start_token, end_token) {
            (T!["("], T![")"]) | (T!["|"], T!["|"]) => {
                ctx.eat();
                break;
            }
            (_, T![","]) => {
                ctx.eat();
            }
            _ => {
                ctx.eat_error(ParseErrorKind::ExpectedToken(T![","]));
                break;
            }
        }
    }

    ctx.finish_node();
}

fn parse_param(ctx: &mut Context) {
    ctx.start_node(PARAM);

    expect_token!(ctx, finish_node, T!["ident"]);

    ctx.finish_node();
}

fn parse_lit(ctx: &mut Context) {
    ctx.start_node(LIT);
    let token = require_token!(ctx, finish_node);

    match token {
        LIT_INT | LIT_FLOAT | LIT_BOOL | LIT_STR | LIT_CHAR => {
            ctx.eat();
        }
        _ => {
            ctx.eat_error(ParseErrorKind::ExpectedOneOfTokens(vec![
                LIT_INT, LIT_FLOAT, LIT_BOOL, LIT_STR, LIT_CHAR,
            ]));
        }
    }

    ctx.finish_node();
}

fn parse_ident(ctx: &mut Context) {
    let token = require_token!(ctx);

    match token {
        IDENT => ctx.eat(),
        _ => ctx.eat_error(ParseErrorKind::ExpectedToken(IDENT)),
    }
}

// Expr =
// | ExprUni
// | ExprBin
// | ExprParen
// | ExprArray
// | ExprIndex
// | ExprObject
// | ExprCall
// | ExprMethodCall
// | ExprAccess
// | ExprClosure
// | ExprIf
// | ExprLoop
// | ExprFor
// | ExprWhile
// | ExprBreak
// | ExprContinue
// | ExprSwitch
// | ExprReturn
// | ExprPath
// | ExprImport
