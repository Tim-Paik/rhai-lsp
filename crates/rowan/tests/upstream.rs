use rhai_rowan::parser::Parser;
use test_case::test_case;

#[test_case(include_str!("../../../rhai/scripts/array.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/assignment.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/comments.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/fibonacci.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/for1.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/for2.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/function_decl1.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/function_decl2.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/function_decl3.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/function_decl4.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/if1.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/if2.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/loop.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/mat_mul.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/module.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/oop.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/op1.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/op2.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/op3.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/primes.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/speed_test.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/string.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/strings_map.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/switch.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/while.rhai"))]
fn parse_rhai_upstream(src: &str) {
    let parse = Parser::new(src).parse_script();
    assert!(parse.errors.is_empty());
    // TODO(tests)
    // insta::assert_snapshot!(format!("{:#?}", parse.into_syntax()));
}
