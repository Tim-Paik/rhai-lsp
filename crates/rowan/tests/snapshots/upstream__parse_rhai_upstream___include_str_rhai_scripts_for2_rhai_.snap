---
source: crates/rowan/tests/upstream.rs
expression: "format!(\"{:#?}\", parse.into_syntax())"

---
RHAI@0..360
  STMT@0..22
    ITEM@0..21
      EXPR@0..21
        EXPR_CONST@0..21
          KW_CONST@0..5 "const"
          WHITESPACE@5..6 " "
          IDENT@6..9 "MAX"
          WHITESPACE@9..10 " "
          OP_ASSIGN@10..11 "="
          EXPR@11..21
            WHITESPACE@11..12 " "
            EXPR_LIT@12..21
              LIT@12..21
                LIT_INT@12..21 "1_000_000"
    PUNCT_SEMI@21..22 ";"
  WHITESPACE@22..24 "\n\n"
  STMT@24..73
    ITEM@24..72
      EXPR@24..72
        EXPR_CALL@24..72
          EXPR@24..29
            EXPR_IDENT@24..29
              IDENT@24..29 "print"
          ARG_LIST@29..72
            PUNCT_PAREN_START@29..30 "("
            EXPR@30..71
              EXPR_LIT@30..71
                LIT@30..71
                  LIT_STR@30..71 "`Iterating an array w ..."
            PUNCT_PAREN_END@71..72 ")"
    PUNCT_SEMI@72..73 ";"
  WHITESPACE@73..75 "\n\n"
  STMT@75..97
    ITEM@75..96
      EXPR@75..96
        EXPR_CALL@75..96
          EXPR@75..80
            EXPR_IDENT@75..80
              IDENT@75..80 "print"
          ARG_LIST@80..96
            PUNCT_PAREN_START@80..81 "("
            EXPR@81..95
              EXPR_LIT@81..95
                LIT@81..95
                  LIT_STR@81..95 "\"Ready... Go!\""
            PUNCT_PAREN_END@95..96 ")"
    PUNCT_SEMI@96..97 ";"
  WHITESPACE@97..99 "\n\n"
  STMT@99..121
    ITEM@99..120
      EXPR@99..120
        EXPR_LET@99..120
          KW_LET@99..102 "let"
          WHITESPACE@102..103 " "
          IDENT@103..106 "now"
          WHITESPACE@106..107 " "
          OP_ASSIGN@107..108 "="
          EXPR@108..120
            EXPR_CALL@108..120
              EXPR@108..118
                WHITESPACE@108..109 " "
                EXPR_IDENT@109..118
                  IDENT@109..118 "timestamp"
              ARG_LIST@118..120
                PUNCT_PAREN_START@118..119 "("
                PUNCT_PAREN_END@119..120 ")"
    PUNCT_SEMI@120..121 ";"
  WHITESPACE@121..123 "\n\n"
  STMT@123..137
    ITEM@123..136
      EXPR@123..136
        EXPR_LET@123..136
          KW_LET@123..126 "let"
          WHITESPACE@126..127 " "
          IDENT@127..131 "list"
          WHITESPACE@131..132 " "
          OP_ASSIGN@132..133 "="
          EXPR@133..136
            WHITESPACE@133..134 " "
            EXPR_ARRAY@134..136
              PUNCT_BRACKET_START@134..135 "["
              PUNCT_BRACKET_END@135..136 "]"
    PUNCT_SEMI@136..137 ";"
  WHITESPACE@137..139 "\n\n"
  STMT@139..185
    ITEM@139..183
      EXPR@139..183
        EXPR_FOR@139..183
          KW_FOR@139..142 "for"
          PAT@142..144
            WHITESPACE@142..143 " "
            PAT_IDENT@143..144
              IDENT@143..144 "i"
          WHITESPACE@144..145 " "
          KW_IN@145..147 "in"
          EXPR@147..162
            EXPR_CALL@147..161
              EXPR@147..153
                WHITESPACE@147..148 " "
                EXPR_IDENT@148..153
                  IDENT@148..153 "range"
              ARG_LIST@153..161
                PUNCT_PAREN_START@153..154 "("
                EXPR@154..155
                  EXPR_LIT@154..155
                    LIT@154..155
                      LIT_INT@154..155 "0"
                PUNCT_COMMA@155..156 ","
                WHITESPACE@156..157 " "
                EXPR@157..160
                  EXPR_IDENT@157..160
                    IDENT@157..160 "MAX"
                PUNCT_PAREN_END@160..161 ")"
            WHITESPACE@161..162 " "
          EXPR_BLOCK@162..183
            PUNCT_BRACE_START@162..163 "{"
            WHITESPACE@163..168 "\n    "
            STMT@168..181
              ITEM@168..180
                EXPR@168..180
                  EXPR_CALL@168..180
                    EXPR@168..177
                      EXPR_BINARY@168..177
                        EXPR@168..172
                          EXPR_IDENT@168..172
                            IDENT@168..172 "list"
                        PUNCT_DOT@172..173 "."
                        EXPR@173..177
                          EXPR_IDENT@173..177
                            IDENT@173..177 "push"
                    ARG_LIST@177..180
                      PUNCT_PAREN_START@177..178 "("
                      EXPR@178..179
                        EXPR_IDENT@178..179
                          IDENT@178..179 "i"
                      PUNCT_PAREN_END@179..180 ")"
              PUNCT_SEMI@180..181 ";"
            WHITESPACE@181..182 "\n"
            PUNCT_BRACE_END@182..183 "}"
    WHITESPACE@183..185 "\n\n"
  STMT@185..227
    ITEM@185..226
      EXPR@185..226
        EXPR_CALL@185..226
          EXPR@185..190
            EXPR_IDENT@185..190
              IDENT@185..190 "print"
          ARG_LIST@190..226
            PUNCT_PAREN_START@190..191 "("
            EXPR@191..225
              EXPR_LIT@191..225
                LIT@191..225
                  LIT_STR@191..225 "`Time = ${now.elapsed ..."
            PUNCT_PAREN_END@225..226 ")"
    PUNCT_SEMI@226..227 ";"
  WHITESPACE@227..229 "\n\n"
  STMT@229..241
    ITEM@229..240
      EXPR@229..240
        EXPR_LET@229..240
          KW_LET@229..232 "let"
          WHITESPACE@232..233 " "
          IDENT@233..236 "sum"
          WHITESPACE@236..237 " "
          OP_ASSIGN@237..238 "="
          EXPR@238..240
            WHITESPACE@238..239 " "
            EXPR_LIT@239..240
              LIT@239..240
                LIT_INT@239..240 "0"
    PUNCT_SEMI@240..241 ";"
  WHITESPACE@241..243 "\n\n"
  STMT@243..276
    ITEM@243..274
      EXPR@243..274
        EXPR_FOR@243..274
          KW_FOR@243..246 "for"
          PAT@246..248
            WHITESPACE@246..247 " "
            PAT_IDENT@247..248
              IDENT@247..248 "i"
          WHITESPACE@248..249 " "
          KW_IN@249..251 "in"
          EXPR@251..257
            WHITESPACE@251..252 " "
            EXPR_IDENT@252..257
              IDENT@252..256 "list"
              WHITESPACE@256..257 " "
          EXPR_BLOCK@257..274
            PUNCT_BRACE_START@257..258 "{"
            WHITESPACE@258..263 "\n    "
            STMT@263..272
              ITEM@263..271
                EXPR@263..271
                  EXPR_BINARY@263..271
                    EXPR@263..267
                      EXPR_IDENT@263..267
                        IDENT@263..266 "sum"
                        WHITESPACE@266..267 " "
                    OP_ADD_ASSIGN@267..269 "+="
                    EXPR@269..271
                      WHITESPACE@269..270 " "
                      EXPR_IDENT@270..271
                        IDENT@270..271 "i"
              PUNCT_SEMI@271..272 ";"
            WHITESPACE@272..273 "\n"
            PUNCT_BRACE_END@273..274 "}"
    WHITESPACE@274..276 "\n\n"
  STMT@276..298
    ITEM@276..297
      EXPR@276..297
        EXPR_CALL@276..297
          EXPR@276..281
            EXPR_IDENT@276..281
              IDENT@276..281 "print"
          ARG_LIST@281..297
            PUNCT_PAREN_START@281..282 "("
            EXPR@282..296
              EXPR_LIT@282..296
                LIT@282..296
                  LIT_STR@282..296 "`Sum = ${sum}`"
            PUNCT_PAREN_END@296..297 ")"
    PUNCT_SEMI@297..298 ";"
  WHITESPACE@298..299 "\n"
  STMT@299..359
    ITEM@299..358
      EXPR@299..358
        EXPR_CALL@299..358
          EXPR@299..304
            EXPR_IDENT@299..304
              IDENT@299..304 "print"
          ARG_LIST@304..358
            PUNCT_PAREN_START@304..305 "("
            EXPR@305..357
              EXPR_LIT@305..357
                LIT@305..357
                  LIT_STR@305..357 "`Finished. Total run  ..."
            PUNCT_PAREN_END@357..358 ")"
    PUNCT_SEMI@358..359 ";"
  WHITESPACE@359..360 "\n"

